### ISSUE TITLE
Redis doesn't disconnect when master goes down during replication

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ashish21590/redis#203]
```
What steps will reproduce the problem?
1. Setup two Redis instances with an amount of data which takes some time 
to replicate (tested with about 2.5GB)
2. Run ./redis-cli slavof server1 port1 on server2
3. Wait for server1 to spawn a process to feed the slave
4. Kill all the Redis processes on server1

What is the expected output? What do you see instead?

The slave should should close the connection to the failed master and start 
accepting the connections from the clients. 

Instead, it seems to try to read the data from the socket using ~100% CPU 
(as usually during the replication). 

What version of the product are you using? On what operating system?

Redis 1.2.5
Debian GNU/Linux 5.0

Please provide any additional information below.

Strace output:

write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0
write(7, ""..., 0)                      = 0
read(6, ""..., 1024)                    = 0

```

Original issue reported on code.google.com by `maciej.l...@gmail.com` on 23 Mar 2010 at 10:21

**Comment by GoogleCodeExporter** [comment]
```
I believe this is related to, but not the same as issue #181

http://code.google.com/p/redis/issues/detail?id=181
```

Original comment by `jzaw...@gmail.com` on 23 Mar 2010 at 3:20

**Comment by GoogleCodeExporter** [comment]
```
I have a branch up @ github that adds 3 new timeout configuration options that 
will enable you to set a 
timeout on the blocking connect and reads used for slave sync:

http://github.com/trickie/redis/commit/f713e5b5e262804f2ab1d3fc602744d1ad5b1db9

That also addresses issue #181

Note: that branch is based on 1.3.x, whilst we developed the patch and have 
tested using 1.2.x. mileage may 
vary, although the same patch for the above commit applies (With some fuzz) 
against 1.2.x.
```

Original comment by `tric...@gmail.com` on 29 Mar 2010 at 12:50

**Comment by GoogleCodeExporter** [comment]
```
The timeout for waiting for the master to finish the background saving could be 
still quite high (as it may take 
some time for the write to disk to complete), and if the master goes down early 
in the process you still need to 
wait potentially a 'long time' before the slave will timeout. 

If you needed to failover to the slave during that time, then there is not much 
you could do. There are two 
options here:

 * make it that a master can stream the db save straight over the network to  a slave (if there is a single slave 
only for example). The timeout of socket ops can then be quite low, and you 
will notice a failed master quite 
soon.
 * potentially make an option where by the slave sync process is handled async like other socket options in the 
event loop, and if you do a 'destructive' client operation, then replication is 
broken off by the slave. The 'be-a-
slave-until-you-have-obviously-been-promoted' option :). That would be quite 
handy for failover using 
keepalived/LVS for example. 

Both of those options sound like they will only be needed by high availability 
setups, but it could be quite 
interesting features.   
```

Original comment by `tric...@gmail.com` on 29 Mar 2010 at 1:02

**Comment by GoogleCodeExporter** [comment]
```
While I have not tested your patch, I've read it and like the idea(s) a lot.  I 
do 
believe it'd fix issue #181, which I reported.

antirez, your thoughts?
```

Original comment by `jzaw...@gmail.com` on 30 Mar 2010 at 10:52

**Comment by GoogleCodeExporter** [comment]
```
The problem of this specific issue is now solved as the slave is now able to 
detect that the connection was closed, but there actually issue #181 still 
open, I just commented there.

Thanks,
Salvatore
```

Original comment by `anti...@gmail.com` on 27 Aug 2010 at 11:15
- Changed state: **Verified**
